spark_utils.py

```diff
@@ -0,0 +1,181 @@
+from itertools import compress
+import pyspark.sql.types as T
+from pyspark.sql.functions import col, explode
+import pyspark.sql.functions as F
+import json
+from twarc import json2csv
+
+def load_schema(path_to_schema):
+    '''Load TweetSets Spark DataFrame schema
+    :param path_to_schema: path to a Spark schema JSON document on disk'''
+    with open(path_to_schema, 'r') as f:
+        schema = json.load(f)
+        return T.StructType.fromJson(schema)
+
+def load_sql(path_to_sql):
+    '''Load Spark SQL code for TweetSets data transform
+    :path_to_sql: path to Spark SQL code on disk'''
+    with open(path_to_sql, 'r') as f:
+        return f.read()
+
+def with_column_index(df):
+    '''Adds an index column to a DataFrame. Used for joining the string representation of the JSON of the full tweet to the parsed JSON of same. (This is necessary to preserve the handling of null fields in the original Twitter API data.
+    
+    :param df: a Spark DataFrame
+    '''
+    # Add a "row" column to the DataFrame (StructField is the pyspark type used for columns, LongType is an integer)
+    new_schema = T.StructType(df.schema.fields + [T.StructField('row', T.LongType(), False),])
+    # The rdd.zipWithIndex allows us to assign a sequential index number across all partitions
+    # Solution here: https://stackoverflow.com/questions/40508489/spark-merge-2-dataframes-by-adding-row-index-number-on-both-dataframes/43746919
+    return df.rdd.zipWithIndex().map(lambda row: row[0] + (row[1],)).toDF(schema=new_schema)
+
+def load_rdd_with_column_index(spark, path_to_tweets):
+    '''Loads a set of JSON tweets as strings and adds a column index. We do this so that the ultimate output in the JSON extract will have the same null fields as the original.
+     
+    :param spark: an initialized SparkSession object
+    :param path_to_dataset: a comma-separated list of JSON files to load'''
```

Comment:
Should this be `:param path_to_tweets:`?

Reply:
Thanks -- fixed.